File: libs/js-shared/base-app-frontend/src/lib/domain/module-definition.ts
import { BaseEntity } from '@processpuzzle/base-entity';
import { RouteDefinition } from './app-definition';

/**
 * Frontend model of the `ModuleDefinition` schema of `base-app-api.yaml`: a lazily-loadable slice of an
 * application, with its own flat route list and its own transloco scope. An {@link ModuleMount} of an
 * `AppDefinition` names one of these by key and mounts its routes under a base path.
 *
 * **What "lazy" honestly means.** A module lazy-loads *metadata* — this definition, its routes, its
 * translations, and the documents and descriptors they name. It does **not** lazy-load widget code:
 * widget components are bundled at compile time and resolved through the frontend registry, so a
 * metadata module cannot split a bundle on its own. The wins are that `AppDefinition` stops being one
 * aggregate that has to be loaded, locked and published atomically, and that a module is the unit of
 * authoring permission and versioning. Nothing here should be justified on bundle size.
 *
 * Unlike the region / route / mount classes this one is a **routable** entity: it has endpoints of its
 * own, so it comes with a service, a store and a facade rather than travelling inside another payload.
 */
export class ModuleDefinition implements BaseEntity {
  /**
   * The contract calls this field `key`; base-entity keys every store, URL and reference on `id`, so
   * {@link ModuleDefinitionMapper} renames it on the way in and out rather than teaching the framework a
   * second identity field. One name for one thing on this side of the mapper — a `key` property beside
   * `id` could only ever disagree with it.
   *
   * Immutable by contract: `ModuleMount.moduleKey` resolves against it, and a rename would silently
   * unmount the module from every app that mounts it.
   */
  id: string;
  name: string;
  translocoId: string | undefined;
  description: string | undefined;
  /**
   * Transloco scope the module's translations load under. Absent means {@link id}, which is the default
   * the contract states; {@link moduleTranslocoScope} is the one place that default is applied.
   */
  translocoScope: string | undefined;
  /**
   * The module's routes, flat and relative to the base path it is mounted at — the same shape and the
   * same derived nesting as `AppDefinition.routes`, and flattened by the same mapper. A module does not
   * mount modules: this is where composition stops.
   */
  routes: RouteDefinition[] | undefined;
  // region server-assigned
  orgKey: string | undefined;
  version: number | undefined;
  createdAt: string | undefined;
  updatedAt: string | undefined;
  // endregion

  constructor(init: Partial<ModuleDefinition> = {}) {
    this.id = init.id ?? '';
    this.name = init.name ?? '';
    this.translocoId = init.translocoId;
    this.description = init.description;
    this.translocoScope = init.translocoScope;
    this.routes = init.routes;
    this.orgKey = init.orgKey;
    this.version = init.version;
    this.createdAt = init.createdAt;
    this.updatedAt = init.updatedAt;
  }
}

/**
 * The scope a module's translations load under, applying the contract's default. Kept as a function
 * rather than resolved in the mapper, so a definition never carries a `translocoScope` the designer did
 * not author — the field stays empty in the form, and the fallback is a rendering decision.
 */
export function moduleTranslocoScope(definition: ModuleDefinition): string {
  return definition.translocoScope?.trim() || definition.id;
}

File: libs/java-shared/base-app-backend/src/main/java/com/processpuzzle/app/domain/ModuleDefinition.java
package com.processpuzzle.app.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Convert;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.IdClass;
import jakarta.persistence.Lob;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;

import java.time.Instant;
import java.util.List;

/**
 * A lazily-loadable slice of an application: its own flat route list, its own transloco scope, its
 * own resource namespace. Identified by ({@code orgKey}, {@code key}).
 *
 * <p>An aggregate of its own rather than a subtree of {@link AppDefinition}, and that is the whole
 * point: it is what stops a large application from being one graph that must be loaded, locked and
 * published atomically, and it makes a module the unit of authoring permission and versioning.
 *
 * <p>A module does <b>not</b> mount modules — {@link ModuleMount} exists only on the application
 * side. Composition therefore recurses exactly one level, by construction rather than by convention.
 *
 * <p>What "lazy" means here is metadata: this definition, its routes, its translations, and the
 * documents and descriptors they name. It is not code splitting — widget components are bundled at
 * compile time and resolved through the frontend registry.
 */
@Entity
@Table(name = "app_modules")
@IdClass(ModuleDefinitionKey.class)
public class ModuleDefinition {

    @Id
    @Column(name = "org_key", length = 63)
    private String orgKey;

    @Id
    @Column(name = "module_key", length = 100)
    private String key;

    @Column(nullable = false, length = 200)
    private String name;

    @Column(name = "transloco_id", length = 200)
    private String translocoId;

    @Column(length = 1000)
    private String description;

    /** Defaults to {@link #key} when unset; see the contract's note on the explicit-alias requirement. */
    @Column(name = "transloco_scope", length = 100)
    private String translocoScope;

    @Lob
    @Column(name = "routes")
    @Convert(converter = RouteListConverter.class)
    private List<AppRoute> routes;

    @Column(nullable = false)
    private long version;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    protected ModuleDefinition() {
        // required by JPA
    }

    public ModuleDefinition(String orgKey, String key, String name) {
        this.orgKey = orgKey;
        this.key = key;
        this.name = name;
        this.version = 1L;
    }

    public void markEdited() {
        this.version++;
    }

    @PrePersist
    void onCreate() {
        Instant now = Instant.now();
        this.createdAt = now;
        this.updatedAt = now;
    }

    @PreUpdate
    void onUpdate() {
        this.updatedAt = Instant.now();
    }

    // region accessors
    public String getOrgKey() {
        return orgKey;
    }

    public String getKey() {
        return key;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getTranslocoId() {
        return translocoId;
    }

    public void setTranslocoId(String translocoId) {
        this.translocoId = translocoId;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    /** The authored scope, or {@link #key} when none was authored. */
    public String getTranslocoScope() {
        return translocoScope == null || translocoScope.isBlank() ? key : translocoScope;
    }

    public void setTranslocoScope(String translocoScope) {
        this.translocoScope = translocoScope;
    }

    public List<AppRoute> getRoutes() {
        return routes == null ? List.of() : routes;
    }

    public void setRoutes(List<AppRoute> routes) {
        this.routes = routes;
    }

    public long getVersion() {
        return version;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public Instant getUpdatedAt() {
        return updatedAt;
    }
    // endregion
}
